chia/wallet/wallet.py

```diff
@@ -265,35 +266,61 @@ async def select_coins(self, amount, exclude: List[Coin] = None) -> Set[Coin]:
-
-        # Use older coins first
-        unspent.sort(reverse=True, key=lambda r: r.coin.amount)
-
         # Try to use coins from the store, if there isn't enough of "unused"
         # coins use change coins that are not confirmed yet
         unconfirmed_removals: Dict[bytes32, Coin] = await self.wallet_state_manager.unconfirmed_removals_for_wallet(
             self.id()
         )
-        for coinrecord in unspent:
-            if sum_value >= amount and len(used_coins) > 0:
-                break
-            if coinrecord.coin.name() in unconfirmed_removals:
+
+        sum_spendable_coins = 0
+        used_coins: Set[Coin] = set()
+        valid_unspent: List[Coin] = []
+
+        for coin_record in unspent:  # remove all the useless coins.
+            if coin_record.coin.name() in unconfirmed_removals:
                 continue
-            if coinrecord.coin in exclude:
+            if coin_record.coin in exclude:
                 continue
-            sum_value += coinrecord.coin.amount
-            used_coins.add(coinrecord.coin)
-            self.log.debug(f"Selected coin: {coinrecord.coin.name()} at height {coinrecord.confirmed_block_height}!")
+            valid_unspent.append(coin_record.coin)
+            sum_spendable_coins += coin_record.coin.amount
 
         # This happens when we couldn't use one of the coins because it's already used
         # but unconfirmed, and we are waiting for the change. (unconfirmed_additions)
-        if sum_value < amount:
+        if sum_spendable_coins < amount:
             raise ValueError(
                 "Can't make this transaction at the moment. Waiting for the change from the previous transaction."
             )
 
+        # Use older coins first
+        valid_unspent.sort(reverse=True, key=lambda r: r.amount)
+
+        # check for exact 1 to 1 coin match.
+        exact_match_coin_record: Optional[Coin] = check_for_exact_match(valid_unspent, amount)
+        if exact_match_coin_record:
+            used_coins.add(exact_match_coin_record)
+            self.log.debug(f"Successfully selected coins: {used_coins}")
+            return used_coins
+
+        # Check for an exact match with all of the coins smaller than the amount.
+        # If we have more, smaller coins than the amount we run the next algorithm.
+        smaller_coin_sum = 0  # coins smaller than target.
+        smaller_coins: Set[Coin] = set()
+        for coin in valid_unspent:
+            if coin.amount < amount:
+                smaller_coin_sum += coin.amount
+                smaller_coins.add(coin)
+        if smaller_coin_sum == amount:
+            used_coins = smaller_coins
+        elif smaller_coin_sum < amount:
+            smallest_coin: Optional[Coin] = find_smallest_coin(valid_unspent, amount)
+            if smallest_coin:
+                used_coins.add(smallest_coin)
+                self.log.debug(f"Selected coin: {smallest_coin.name()}")
+        elif smaller_coin_sum > amount:
```

Comment:
This can be an else instead of elif. There are no other remaining cases, the only cases are `=`. `<`, and `>`